### ISSUE TITLE
nightly 12/28: 5-15-NFS-Datastore panics on VCH delete 

### ISSUE THREAD

**Issue by AngieCris** [vmware/vic#7016]
In test cleanup, VCH delete panics and CDE.

`vic-machine.log` contains a stack trace:
```
Dec 29 2017 03:00:02.143-06:00 INFO  op=10546.1: ### Removing VCH ####
Dec 29 2017 03:30:02.143-06:00 ERROR op=10546.1: vic/pkg/trace.(*Operation).Err: vic-machine-linux error: context deadline exceeded
vic/cmd/vic-machine/delete.(*Uninstall).Run:92 vic-machine-linux
vic/cmd/vic-machine/common.NewOperation:27 vic-machine-linux
Dec 29 2017 03:30:02.145-06:00 ERROR op=10546.1: vic/pkg/trace.(*Operation).Err: vic-machine-linux error: context deadline exceeded
......
Dec 29 2017 03:30:02.145-06:00 ERROR op=10546.1: --------------------
Dec 29 2017 03:30:02.145-06:00 ERROR op=10546.1: vic-machine-linux delete failed: Delete timed out: use --timeout to add more time

goroutine 1 [running]:
runtime/debug.Stack(0xc42001c5f0, 0xfcc546, 0x2f)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x79
main.main.func1(0xc420067520, 0xc420084288)
	/go/src/github.com/vmware/vic/cmd/vic-machine/main.go:160 +0x298
panic(0xda3b60, 0x1885f90)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/vmware/vic/pkg/vsphere/session.(*Session).Populate(0xc420206a00, 0x189bda0, 0xc4201edb80, 0xd, 0x1, 0x0)
	/go/src/github.com/vmware/vic/pkg/vsphere/session/session.go:358 +0xf8a
github.com/vmware/vic/lib/install/validate.NewValidator(0x189bda0, 0xc4201edb80, 0xc4201df400, 0x0, 0x0, 0x0)
	/go/src/github.com/vmware/vic/lib/install/validate/validator.go:159 +0x978
github.com/vmware/vic/cmd/vic-machine/delete.(*Uninstall).Run(0xc4201d58f0, 0xc42017db80, 0x1895de0, 0xc4204766e0)
	/go/src/github.com/vmware/vic/cmd/vic-machine/delete/delete.go:112 +0x2d6
github.com/vmware/vic/cmd/vic-machine/delete.(*Uninstall).Run-fm(0xc42017db80, 0x1010101010100, 0xc42017db80)
	/go/src/github.com/vmware/vic/cmd/vic-machine/main.go:75 +0x34
github.com/vmware/vic/vendor/gopkg.in/urfave/cli%2ev1.HandleAction(0xd56cc0, 0xc4201d5940, 0xc42017db80, 0xc42006b100, 0x0)
	/go/src/github.com/vmware/vic/vendor/gopkg.in/urfave/cli.v1/app.go:485 +0xd4
github.com/vmware/vic/vendor/gopkg.in/urfave/cli%2ev1.Command.Run(0xf8ef69, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0xfbf004, 0x23, 0x0, ...)
	/go/src/github.com/vmware/vic/vendor/gopkg.in/urfave/cli.v1/command.go:193 +0xb72
github.com/vmware/vic/vendor/gopkg.in/urfave/cli%2ev1.(*App).Run(0xc420067520, 0xc420088000, 0xa, 0xa, 0x0, 0x0)
	/go/src/github.com/vmware/vic/vendor/gopkg.in/urfave/cli.v1/app.go:250 +0x7d0
main.main()
	/go/src/github.com/vmware/vic/cmd/vic-machine/main.go:165 +0x2233
```

From the stack trace, the panic happens during validation session populate, where it grabs VM folder from datacenter using govmomi:
```
	if s.Datacenter != nil {
		folders, err := s.Datacenter.Folders(op)
		if err != nil {
			errs = append(errs, fmt.Sprintf("Failure finding folders (%s): %s", s.DatacenterPath, err.Error()))
		} else {
			op.Debugf("Cached folders: %s", s.DatacenterPath)
		}
		s.VMFolder = folders.VmFolder
	}
```
In line `s.VMFolder = folders.VmFolder`.
https://github.com/vmware/vic/blob/master/pkg/vsphere/session/session.go#L358

Log: [5-15-NFS-Datastore.zip](https://github.com/vmware/vic/files/1598194/5-15-NFS-Datastore.zip)

Cc: @dougm @hickeng

**Comment by hickeng** [comment]
There are two issues I see here:
1. we should not be using `folders` in the fragment from `session.go` if there was an error returned - it's almost certain to be incomplete or nil.
2. we will ignore the error in Validator because we do not expect session.Populate to complete cleanly - we simply should not be using session.Populate in it's current form, it's not suited for our use case. The improper use of session is a known techdebt item (#1387)

**Comment by AngieCris** [comment]
I can only think of one possible cause for this panic:
`s.Datacenter.Folders(op)` actually failed and returned an error, which causes the datastore folder object returned to be incomplete.

However currently it's hard to say if this is the root cause, because:
1. Any error from `s.Datacenter.Folders(op)` is logged at debug level, and the robot keyword we use to delete VCH (`Run Secrete VIC Delete Command`) does not turn on debug
2. This step `s.Datacenter.Folders(op)` is from validator session populate, but any failure/error returned from session populate is ignored by validator, so even if an error is returned to the caller, we still have no way to see it/track it down

The reason why 2. occurs is that, as @hickeng explains above, we abuse the use of session here and we never expect session to come out cleanly.
So without the relevant error info it's hard to actually fix the nightly issue from the root.

Things we could do to improve the situation and gets us closer to actually fixing it are:
1. Enable debug in VCH delete robot keyword
2. If `s.Datacenter.Folders(op)` fails:
    1. Do we want to keep deleting the VCH or do we want to bail out?
    2. Is a retry pattern around `s.Datacenter.Folders(op)` suitable for the circumstances? 

@hickeng

**Comment by AngieCris** [comment]
```
	if s.Datacenter != nil {
		folders, err := s.Datacenter.Folders(op)
		if err != nil {
			errs = append(errs, fmt.Sprintf("Failure finding folders (%s): %s", s.DatacenterPath, err.Error()))
		} else {
			op.Debugf("Cached folders: %s", s.DatacenterPath)
		}
		s.VMFolder = folders.VmFolder
	}
```

The above piece of code that fails in `pkg/vsphere/session/session.go` function `session.Populate` is semantically wrong. If `s.Datacenter.Folders(op)` returns an error, the folder returned is nil, then we shouldn't be setting `session.VMFolder`. In this case, vic-machine create process should fail out. 
But currently we ignore any error returned by session populate and continue on creating the VCH. This should be changed.

The root cause of why `s.Datacenter.Folders(op)` returns an error is probably related to #7450. After #7450 merges in, we can wait and see if there's future reoccurrence of this nightly failure.

**Comment by AngieCris** [comment]
Any further changes on validator logic to prevent nil `validator.session.VMFolder` from validator session populate might potentially cause issues for multi-datacenter vic-machine operations. 

We'll wait until vch inventory folder support #773  checks in and revisit then.
Moving back to To-Do.

**Comment by renmaosheng** [comment]
https://vic-jenkins.eng.vmware.com/view/Scenario/job/vic-master-scenario/102/robot/report/log.html#s1-s4-s14, the test continuously pass.  please file a new issue if we hit it again.
